ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/State/Types.hs

```diff
@@ -125,10 +132,92 @@ newtype CrossEraForecaster state view x y = CrossEraForecaster {
         -> SlotNo   -- 'SlotNo' we're constructing a forecast for
-        -> state x
+        -> state x EmptyMK
         -> Except OutsideForecastRange (view y)
     }
 
+-- | Translate a 'LedgerState' across an era transition.
+newtype TranslateLedgerState x y = TranslateLedgerState {
+  -- | How to translate a 'LedgerState' during the era transition.
+  --
+  -- When translating between eras, it can be the case that values are modified,
+  -- thus requiring this to be a @DiffMK@ on the return type. If no tables are
+  -- populated, normally this will be filled with @emptyLedgerTables@.
+  --
+  -- To make a clear example, in the context of Cardano, there are currently two
+  -- cases in which this is of vital importance: Byron->Shelley and
+  -- Shelley->Allegra.
+  --
+  -- On Byron->Shelley we basically dump the whole UTxO set as insertions
+  -- because the LedgerTables only exist for Shelley blocks.
+  --
+  -- On Shelley->Allegra, there were a bunch of UTxOs that were moved around,
+  -- related to the AVVMs. In particular they were deleted and included in the
+  -- reserves. See the code that performs the translation Shelley->Allegra for
+  -- more information.
+    translateLedgerStateWith ::
+         EpochNo
+      -> LedgerState x EmptyMK
+      -> LedgerState y DiffMK
+  }
+
+-- | Transate a 'LedgerTables' across an era transition.
+data TranslateLedgerTables x y = TranslateLedgerTables {
+    -- | Translate a 'TxIn' across an era transition.
+    --
+    -- See 'translateLedgerTablesWith'.
+    translateTxInWith  :: !(Key (LedgerState x) -> Key (LedgerState y))
+
+    -- | Translate a 'TxOut' across an era transition.
+    --
+    -- See 'translateLedgerTablesWith'.
+  , translateTxOutWith :: !(Value (LedgerState x) -> Value (LedgerState y))
+  }
+
+newtype TranslateTxIn x y = TranslateTxIn (Key (LedgerState x) -> Key (LedgerState y))
+
+newtype TranslateTxOut x y = TranslateTxOut (Value (LedgerState x) -> Value (LedgerState y))
+
+-- | Translate a 'LedgerTables' across an era transition.
+--
+-- To translate 'LedgerTable's, it's sufficient to know how to translate
+-- 'TxIn's and 'TxOut's. Use 'translateLedgerTablesWith' to translate
+-- 'LedgerTable's using 'translateTxInWith' and 'translateTxOutWith'.
+--
+-- This is a rather technical subtlety. When performing a ledger state
+-- translation, the provided input ledger state will be initially populated with
+-- a @emptyLedgerTables@. This step is required so that the operation provided
+-- to 'Telescope.extend' is an automorphism.
+--
+-- If we only extend by one era, this function is a no-op, as the input will be
+-- empty ledger states. However, if we extend across multiple eras, previous
+-- eras might populate tables thus creating values that now need to be
+-- translated to newer eras. This function fills that hole and allows us to
+-- promote tables from one era into tables from the next era.
+--
+-- TODO(jdral): this is not optimal. If either 'translateTxInWith' or
```

Comment:
reifying a TODO about skipping traversals if the key/value translations are `id`

Reply:
Might also consider combining the two traversals into a single traversal. And maybe also something about monotonicity.

Reply:
This function is only used in an InPairs.extend so it won't even be called with an 'id'. Noting that down on a comment.